R/Spectra-functions.R

```diff
@@ -623,4 +588,91 @@ joinSpectraData <- function(x, y,
 #' @rdname Spectra
 processingLog <- function(x) {
     x@processing
-}
\ No newline at end of file
+}
+
+#' @export
+#'
+#' @rdname Spectra
+estimatePrecursorIntensity <- function(x, ppm = 10, tolerance = 0,
+                                       method = c("previous", "interpolation"),
+                                       msLevel. = 2L, f = dataOrigin(x),
+                                       BPPARAM = bpparam()) {
+    if (is.factor(f))
+        f <- as.character(f)
+    f <- factor(f, levels = unique(f))
+    unlist(bplapply(split(x, f), FUN = .estimate_precursor_intensity, ppm = ppm,
+                    tolerance = tolerance, method = method, msLevel = msLevel.,
+                    BPPARAM = BPPARAM), use.names = FALSE)
+}
+
+#' estimate precursor intensities based on MS1 peak intensity. This function
+#' assumes that `x` is a `Spectra` with data **from a single file/sample**.
+#'
+#' @author Johannes Rainer
+#'
+#' @importFrom stats approx
+#'
+#' @noRd
+.estimate_precursor_intensity <- function(x, ppm = 10, tolerance = 0,
+                                          method = c("previous",
+                                                     "interpolation"),
+                                          msLevel = 2L) {
+    method <- match.arg(method)
+    if (any(msLevel > 2))
+        warning("Estimation of precursor intensities for MS levels > 2 is ",
+                "not yet validated")
+    if (!length(x))
+        return(numeric())
+    if (is.unsorted(rtime(x))) {
+        idx <- order(rtime(x))
+        x <- x[idx]
+    } else idx <- integer()
+    pmz <- precursorMz(x)
+    pmi <- rep(NA_real_, length(pmz))
+    idx_ms2 <- which(!is.na(pmz) & msLevel(x) == msLevel)
+    x_ms1 <- filterMsLevel(x, msLevel = msLevel - 1L)
+    pks <- .peaksapply(x_ms1)
+    for (i in idx_ms2) {
+        rt_i <- rtime(x[i])
+        before_idx <- which(rtime(x_ms1) < rt_i)
```

Comment:
Do we use something like this often? If that is the case I am going to implement a `which.first`/`which.last` function in `MsCoreUtils`? What do you think?

Reply:
Hm, actually, getting the index of the last (x) retention time(s) before and the first (x) after might indeed be something we could use more frequently.

Reply:
That's a nice small function. I am going to take that one: https://github.com/rformassspectrometry/MsCoreUtils/issues/85 in the next days.

Reply:
cool!